src/twisted/test/test_twistd.py

```diff
@@ -1545,11 +1545,13 @@ def test_getLogObserverFile(self):
         sut = app.AppLogger({"logfile": filename})
 
         observer = sut._getLogObserver()
-        self.addCleanup(logger.globalLogPublisher.removeObserver, observer)
+        self.addCleanup(observer._outFile.close)
 
         self.assertEqual(len(logFiles), 1)
         self.assertEqual(logFiles[0].path, os.path.abspath(filename))
 
+        logger.globalLogPublisher.removeObserver(observer)
```

Comment:
Is there a reason to not register both as cleanup activities?  As is, won't a failed assertion result in leaving the observer in place?  In the changes below, the observer doesn't get removed in any case anymore.  The inconsistency isn't obviously good.

(not that I know anything about twisted logging...)

Reply:
Seems correct to me.

Reply:
(I mean @altendky 's comment seems correct to me.)

Reply:
Since it isn't just my naiveté, I'll formally elevate this to 'request changes'.

Reply:
Good question.
Sorry for my bad code over here.

`logger.globalLogPublisher.removeObserver(observer)` does nothing in this case and should be removed.

I was reading the `AppLogger._getLogObserver()` implementation
 https://github.com/twisted/twisted/blob/trunk/src/twisted/application/app.py#L228

I saw the file is opened in `logfile.LogFile.fromFullPath(self._logfilename)` 
and then I saw that that remove observer is called in `AppLogger.stop()` and I think that this is how you can clean/close what was open by `_getLogObserver`

But I was wrong.
The stop doesn't closes the file.

I tried to read the code for the logging framework, but I am always reading it after 10PM and I can't make sense of it.

I have no idea how this logging API should be used.

This test was leaving open files for a long time, is just that it was not noticed since we haven't executed parallel tests or garbage collection checks.

You can run this on trunk
```
$ python3 -Wall -m twisted.trial --force-gc twisted.test.test_twistd.AppLoggerTests

    test_getLogObserverFile ...                                            [OK]
/home/adi/chevah/twisted/src/twisted/trial/_asyncrunner.py:141: ResourceWarning: unclosed file <_io.FileIO name='/home/adi/chevah/twisted/_trial_temp/twisted.test.test_twistd/AppLoggerTests/test_getLogObserverFile/q_avmsk1/temp' mode='rb+' closefd=True>
    test_getLogObserverStdout ...                                          [OK]

```

and then run the same test on this branch and you will not see the ResourceWarning.


I was fooled by calling the test direclty and not seeing the warning and thinking that it was fixed...

```
$ python3 -Wall -m twisted.trial --force-gc twisted.test.test_twistd.AppLoggerTests.test_getLogObserverFile
```

this is what you happend when you try to skip some TDD steps

Reply:
Ok, so `logger.globalLogPublisher.removeObserver()` is irrelevant because the observer is never getting added to the global publisher to begin with...  right?  So then yeah, I finally get that it doesn't matter.

https://github.com/twisted/twisted/blob/6fc21a15d345d32328e2df622870ee1e98868f97/src/twisted/application/app.py#L228-L237

It crossed my mind that there's some 'hazard' of the file being `stdout`, but we are explicitly setting the file in each of these cases so it would probably be overly cautious to worry about a change that would result in ...  meh...  yeah.  Looks good.  Thanks.

Reply:
even if the observer would have been added to the global log publisher, I think that the API call for remove will just "stop forwarding logs to this observer". It will not take care of observer close and cleanup.